util/misc.go

```diff
@@ -531,3 +543,62 @@ func GetLocalIP() string {
 	}
 	return ""
 }
+
+func createTLSCertificates() error {
+
+	privkey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return err
+	}
+
+	notBefore := time.Now()
+	notAfter := notBefore.Add(90 * 24 * time.Hour)
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		NotBefore:    notBefore,
+		NotAfter:     notAfter,
+		DNSNames:     []string{hostname},
+	}
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, &privkey.PublicKey, privkey)
+	if err != nil {
+		return err
+	}
+
+	tempStoragePath := config.GetGlobalConfig().TempStoragePath
+	certOut, err := os.Create(tempStoragePath + "/cert.pem")
+	if err != nil {
+		return err
+	}
+	if err := pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes}); err != nil {
+		return err
+	}
+
+	if err := certOut.Close(); err != nil {
+		return err
+	}
+
+	keyOut, err := os.OpenFile(tempStoragePath+"/key.pem", os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return err
+	}
+
+	privBytes, err := x509.MarshalPKCS8PrivateKey(privkey)
+	if err != nil {
+		return err
+	}
+
+	if err := pem.Encode(keyOut, &pem.Block{Type: "PRIVATE KEY", Bytes: privBytes}); err != nil {
+		return err
+	}
+
+	if err := keyOut.Close(); err != nil {
```

Comment:
I think you can use `defer keyOut.Close();` directly after `os.OpenFile` to simplify the handling.

Reply:
I don't think that would return the error in the same way and might hide the error.

Reply:
I think hiding a potential error from os.Close() is better than keeping the file opened (currently not sure what go's garbage collection does with open files that don't have any reference to them any longer...)
I think the only possible error from os.Close is that it already been closed:
https://golang.org/pkg/os/#File.Close
So instead of adding os.Close to all return's after os.OpenFile(), I think using `defer keyOut.Close()` is the better option.

Reply:
Yes, I agree.

Reply:
Looks like this isn't allowed.

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb_ghpr_check/detail/tidb_ghpr_check/82829/pipeline
```
[2021-04-20T09:05:25.745Z] + GOPATH=/home/jenkins/agent/workspace/tidb_ghpr_check/go
[2021-04-20T09:05:25.745Z] + make check
[2021-04-20T09:05:25.745Z] gofmt (simplify)
[2021-04-20T09:05:40.610Z] cd tools/check; \
[2021-04-20T09:05:40.610Z] GO111MODULE=on go build -o ../bin/errcheck github.com/kisielk/errcheck
[2021-04-20T09:05:40.610Z] errcheck
[2021-04-20T09:05:48.726Z] util/misc.go:577:21:	defer certOut.Close()
[2021-04-20T09:05:48.726Z] util/misc.go:586:20:	defer keyOut.Close()
[2021-04-20T09:05:48.726Z] make: *** [errcheck] Error 1
script returned exit code 2
```

Reply:
:( Seems like there are several ways to handle it, non really nice and straight forward:
https://www.joeshaw.org/dont-defer-close-on-writable-files/
I personally prefer the anonymous defer function like:
```
    defer func() {
        cerr := f.Close()
        if err == nil {
            err = cerr // propagate the error if no error is already set!
        }
    }()
  ```

Maybe with the extra `return os.Close() // Explicitly return the possible error from os.Close()` but now I almost starts an bike shed discussion...

Reply:
Going for the original code for now